src/rankcorr.jl

```diff
@@ -28,71 +28,76 @@ corspearman(X::RealMatrix) = (Z = mapslices(tiedrank, X, dims=1); cor(Z, Z))
 
 
 #######################################
-#
+# 
 #   Kendall correlation
-#
+# 
 #######################################
 
-# Knight JASA (1966)
-
-function corkendall!(x::RealVector, y::RealVector)
+# Knight, William R. “A Computer Method for Calculating Kendall's Tau with Ungrouped Data.”
+# Journal of the American Statistical Association, vol. 61, no. 314, 1966, pp. 436–439.
+# JSTOR, www.jstor.org/stable/2282833. Accessed 15 Jan. 2021.
+function corkendall!(x::RealVector, y::RealVector, permx=sortperm(x))
     if any(isnan, x) || any(isnan, y) return NaN end
     n = length(x)
     if n != length(y) error("Vectors must have same length") end
 
     # Initial sorting
-    pm = sortperm(y)
-    x[:] = x[pm]
-    y[:] = y[pm]
-    pm[:] = sortperm(x)
-    x[:] = x[pm]
-
-    # Counting ties in x and y
-    iT = 1
-    nT = 0
-    iU = 1
-    nU = 0
-    for i = 2:n
-        if x[i] == x[i-1]
-            iT += 1
-        else
-            nT += iT*(iT - 1)
-            iT = 1
-        end
-        if y[i] == y[i-1]
-            iU += 1
-        else
-            nU += iU*(iU - 1)
-            iU = 1
+    x[:] = x[permx]
+    y[:] = y[permx]
+
+    npairs = float(n) * (n - 1) / 2
+    # ntiesx, ntiesy, ndoubleties are floats to avoid overflows on 32bit
+    ntiesx, ntiesy, ndoubleties, k, nswaps = 0.0, 0.0, 0.0, 0, 0
```

Comment:
AFAICT overflow can happen also on 64-bit with vectors larger than `sqrt(typemax(Int))`, right? `npairs` is also a float for this reason so the comment should mention it.

More fundamentally, are we sure that even `Float64` is large enough to avoid loss of precision in extreme cases? For example, imagine a vector with 1G entries (taking 8GB in RAM): IIUC it can happen that we will try to compute a number of the order of `1e9^2`, and yet `1e9^2 - (1e9^2+1) == 0` so everything is going to collapse in that case right? Maybe using `widen(Int)`/`widen(0)` would be safer, so that we use `Int128` on 64-bit machines?

Reply:
Yes you are right. The issue was a nagging worry at the time, and I should have asked your advice about it.  There are (at least) two pitfalls that need to be avoided:
1. The type we use doesn't allow us to "count high enough" because of the `(x+1) == x` problem. I think a Float64 "stops working as a counter" at 2^53. So we must have `npairs <2^53` which in turn means we must have `length(x) < 2^27`. So that means we should abandon Float64 and go with your suggestion of `widen(Int)` with counters initialised to `widen(0)`.
2. The risk that the last line of function `corkendall` fails due to overflow problems.

I will make some changes and re-submit.

Reply:
OK, that's submitted now, but tests are failing, and I think I see why. Was a mistake to initialise `k` to wide. Weird thing is that tests passed on my PC (Windows 64bit)

Reply:
Ah I see my prior comment looks a bit cryptic. I'd previously written a long comment basically agreeing with your analysis, and somehow that got lost. Sorry.